app/src/db/migrations/20220627000000_002-metadata-tags.js

```diff
@@ -0,0 +1,122 @@
+    .then(() => knex.schema.createTable('metadata', table => {
+      table.specificType(
+        'id',
+        'integer GENERATED ALWAYS AS IDENTITY PRIMARY KEY'
+      );
+      table.string('key').index();
+      table.string('value').index();
+      table.unique(['key', 'value']);
+
+    }))
+
+    // create version_metadata table
+    .then(() => knex.schema.createTable('version_metadata', table => {
+      table.uuid('versionId').notNullable().references('id').inTable('version');
+      table.integer('metadataId').notNullable().references('id').inTable('metadata');
+      stamps(knex, table);
+      table.primary(['versionId', 'metadataId']);
+    }))
+
+    // create tag table
+    .then(() => knex.schema.createTable('tag', table => {
+      table.specificType(
+        'id',
+        'integer GENERATED ALWAYS AS IDENTITY PRIMARY KEY'
+      );
+      table.string('key').index();
+      table.string('value').index();
+      table.unique(['key', 'value']);
+    }))
+
+    // create version_tag table
+    .then(() => knex.schema.createTable('version_tag', table => {
+      table.uuid('versionId').notNullable().references('id').inTable('version');
+      table.integer('tagId').notNullable().references('id').inTable('tag');
+      stamps(knex, table);
+      table.primary(['versionId', 'tagId']);
+    }))
+
+
+    // Create metadata audit trigger
+    .then(() => knex.schema.raw(`CREATE TRIGGER audit_version_metadata_trigger
+    AFTER UPDATE OR DELETE ON version_metadata
+    FOR EACH ROW EXECUTE PROCEDURE audit.if_modified_func();`))
+    // Create tag audit trigger
+    .then(() => knex.schema.raw(`CREATE TRIGGER audit_version_tag_trigger
+    AFTER UPDATE OR DELETE ON version_tag
+    FOR EACH ROW EXECUTE PROCEDURE audit.if_modified_func();`))
+
+  /**
+   * for each version, move originalName and objectId to records in metadata table
+   * and create joining version_metadata records
+   */
+    .then(() => knex('version').where('deleteMarker', false))
+    .then(async (rows) => {
+      const versions = rows.map(v => ({
+        value: v.originalName,
+        versionId: v.id,
+        objectId: v.objectId
+      }));
+      let promises = versions.map(async (row) => {
+        // insert into metadata table, returning [metadata.id]
+        const insertIdArray = await knex('metadata').insert([
+          { key: 'name', value: row.value },
+          { key: 'id', value: row.objectId }
+        ]).onConflict(['key', 'value'])
+          .merge()
+          .returning('id');
+        // add joining records
+        return knex('version_metadata').insert([
+          { versionId: row.versionId, metadataId: insertIdArray[0].id },
+          { versionId: row.versionId, metadataId: insertIdArray[1].id }
+        ]);
+      });
+      await Promise.all(promises);
+    })
+
+    // remove column originalName from version table
+    .then(() => knex.schema.alterTable('version', table => {
+      table.dropColumn('originalName');
+    }));
```

Comment:
I'm not convinced we should remove the originalName column just yet -> likely need further discussion.
If we do go through with removing the originalName column, then there will need to be minimal associated changes to our control layer to make sure that the metadata we're recording is getting written into the new models; if we merged this PR as is now, it would likely fail or error.